T4958: ocserv: openconnect: Add RADIUS accounting support

PeppyH opened the pull request:
Adds CLI configuration options to configure RADIUS accounting for OpenConnect VPN sessions. This functionality cannot be used outside of the RADIUS OpenConnect VPN authentication mode

<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Adds support for configuring OpenConnect to send accounting packets to a (optionally separate) RADIUS server when also using RADIUS for OpenConnect authentication.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T4958

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
ocserv

## Proposed changes
<!--- Describe your changes in detail -->
#### Additions
CLI configuration node "accounting" under `vpn` -> `openconnect` for configuring RADIUS accounting server

New accounting entries have been added to the `ocserv_config` and `radius_conf` templates.

Verification code in the `verify` function of `vpn_openconnect.py` for validating that RADIUS accounting cannot be used without also configuring openconnect to use radius for authentication.

### Refactors
Refactors the existing templates for `radius-server-ipv4.xml.i` splitting it into two seperate templates: `radius-auth-server-ipv4.xml.i` and `radius-server-acct-port.xml.i` to allow for separate default port values. Additionally the existing usages of the `radius-server-ipv4.xml.i` have been refactored to reflect the name change to `radius-auth-server-ipv4.xml.i`

The structure of the config data supplied to the `radius_conf` template in `vpn_openconnect.py` has been modified to allow for distinguishing between authentication and accounting. The template itself has been refactored to reflect this. Additionally when both radius accounting and authentication is configured, the `radius_servers` config file is populated using a merged dictionary of both authentication and accounting servers, so that if a separate accounting server used is to the authentication server, both server entries will be populated in the `radius_servers` config.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
As a prerequisite configure a RADIUS authentication server for openconnect.
```
set vpn openconnect accounting server {server ip} port {server port}
set vpn openconnect accounting server {server ip} key {radius shared secret}
set vpn openconnect mode radius
commit
save
```

### ocserv config
```
sudo cat /run/ocserv/ocserv.conf
```
And observe that above the `auth` entry there is a new entry for `acct` e.g.:
```
acct = "radius [config=/run/ocserv/radiusclient.conf]"
```

### radiusclient config
```
sudo cat /run/ocserv/radiusclient.conf
```
And observe the accounting server entries as so
```
#### Accounting
acctserver {yourip}:{yourport}
```

### radius_servers config
This should output a server and shared secret as per CLI
```
sudo cat /run/ocserv/radius_servers
```

### Radius Server Accounting
To observe the RADIUS accounting, connect to the OpenConnect server using credentials configured on your RADIUS server. If you have configured a SQL database the RADIUS server you can observe the logs by querying the radacct table in the radius database using a default setup as per [freeradius wiki](https://wiki.freeradius.org/guide/sql-howto). Alternatively you can probably just observe the radius accounting packets coming in by tailing logs if you start freradius in debug mode.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly


If this PR is accepted I'll submit a PR to the documentation repo to add configuration examples. Additionally I may write a blogpost to include configuring this from not only a VyOS perspective but setting up a FREERADIUS server with SQL accounting and provide examples of retrieving some statistics from the database.

Commit:
T4958: ocserv: openconnect: adds support for configuring RADIUS accounting

Adds CLI configuration options to configure RADIUS accounting for OpenConnect VPN sessions. This functionality cannot be used outside of the RADIUS OpenConnect VPN authentication mode

sarthurdev commented on data/templates/ocserv/ocserv_config.j2:
```diff
@@ -10,6 +10,10 @@ udp-port = {{ listen_ports.udp }}
 run-as-user = nobody
 run-as-group = daemon
 
+{% if "radius" in accounting.mode %}
```
Would be best to use vyos_defined here, e.g: `if accounting.mode.radius is vyos_defined`

sarthurdev commented on data/templates/ocserv/radius_conf.j2:
```diff
@@ -1,20 +1,38 @@
-authserver {{ srv }}
+
+#### Accounting
+{% if "radius" in accounting['mode'] %}
```
Would be best to use vyos_defined here, e.g: `if accounting.mode.radius is vyos_defined`

sarthurdev commented on data/templates/ocserv/radius_conf.j2:
```diff
@@ -1,20 +1,38 @@
+{% endif %}
+
+#### Authentication
+{% if "radius" in authentication['mode'] %}
```
Would be best to use vyos_defined here, e.g: `if authentication.mode.radius is vyos_defined`

sarthurdev commented on data/templates/ocserv/radius_conf.j2:
```diff
@@ -1,20 +1,38 @@
+{%     for acctsrv in accounting['radius']['server'] %}
+{%         if not "disable" in accounting['radius']['server'][acctsrv] %}
+{%             if "port" in accounting['radius']['server'][acctsrv] %}
+acctserver {{ acctsrv }}:{{ accounting['radius']['server'][acctsrv]['port'] }}
```
This could be shortened to:
```
{% for acctsrv, srv_conf in accounting.radius.server.items() if 'disable' not in srv_conf %}
{%     if srv_conf.port is vyos_defined %}
acctserver {{ acctsrv }}:{{ srv_conf.port }}
```

sarthurdev commented on data/templates/ocserv/radius_conf.j2:
```diff
@@ -1,20 +1,38 @@
+{%     for authsrv in authentication['radius']['server'] %}
+{%         if not "disable" in authentication['radius']['server'][authsrv] %}
+{%             if "port" in authentication['radius']['server'][authsrv] %}
+authserver {{ authsrv }}:{{ authentication['radius']['server'][authsrv]['port'] }}
```
This could also be shortened like above comment.

sarthurdev commented on src/conf_mode/vpn_openconnect.py:
```diff
@@ -166,10 +174,20 @@ def generate(ocserv):
-        render(radius_cfg, 'ocserv/radius_conf.j2', ocserv["authentication"]["radius"])
-        # Render radius servers
-        render(radius_servers, 'ocserv/radius_servers.j2', ocserv["authentication"]["radius"])
+        if "accounting" in ocserv and "mode" in ocserv["accounting"] and "radius" in ocserv["accounting"]["mode"]:
```
Can be simplified using vyos.util.dict_search: `if dict_search(ocserv, 'accounting.mode.radius'):`

sarthurdev commented on src/conf_mode/vpn_openconnect.py:
```diff
@@ -166,10 +174,20 @@ def generate(ocserv):
+            # Merge the accounting and authentication servers into a single dictionary
+            render(radius_servers, 'ocserv/radius_servers.j2', {'server': merged_servers})
+        else:
+            acct_and_auth_config = {'accounting': {'mode': ''}, 'authentication': ocserv['authentication']}
```
Using `vyos_defined` checks in Jinja templates means you won't need to set accounting.mode to an empty string here.

PeppyH committed:
T4958: ocserv: openconnect: refactor RADIUS accounting support

PeppyH commented:
Thanks for the feedback [sarthurdev](https://github.com/sarthurdev) I've pushed new changes based in line with your comments. Please advise if you'd like these two commits squashed and i'll do so

PeppyH committed:
openconnect: T4955: Removed wrong acctserver in radiusclient.conf

Removes port key from accounting server merged config dictionary.

PeppyH commented:
I rebased the PR branch from current to test compatibility with [T4955](https://github.com/vyos/vyos-1x/pull/1785) - I found no compatibility issues but found the issues addressed T4955 presented in the accounting configs added also, so I have added a new commit which uses the same logic to solve the issue.

I've ammended the original comment to reflect this
